File: 0813-all-paths-from-source-to-target/0813-all-paths-from-source-to-target.cs
public class Solution {
    public IList<IList<int>> AllPathsSourceTarget(int[][] graph) {
        var ans = new List<IList<int>>();
        var path = new List<int>();
        int n = graph.Length;
        var vis = new bool[n];

        void dfs (int node) {
            vis[node] = true;
            path.Add(node);
            if (node == n - 1)
                ans.Add(new List<int>(path));
            
            foreach (var neigh in graph[node]) {
                if (!vis[neigh])
                    dfs (neigh);
            }

            vis[node] = false;
            path.RemoveAt(path.Count - 1);
        }

        dfs (0);

        return ans;
    }
}

File: 0813-all-paths-from-source-to-target/0813-all-paths-from-source-to-target.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        n = len(graph)
        ans = []
        path = []
        vis = [False for _ in range(n)]

        def dfs (node : int) -> None:
            vis[node] = True
            path.append(node)
            if node == n - 1:
                ans.append([x for x in path])
            
            for neigh in graph[node]:
                if not vis[neigh]:
                    dfs (neigh)
            
            vis[node] = False
            path.pop()
        
        dfs (0)
        return ans